### PR TITLE
Source generators produce multiline strings

### LINKED ISSUE #215: sourceLineRanges is inconsistent with its spec
`sourceLineRanges` and its spec differ in their opinions of carriage returns: https://travis-ci.org/github/semantic/builds/566542927#L1680-L1698

This wasn’t discovered until quite recently, so the test’s distribution clearly needs some shaping.

### PR DESCRIPTION
**Pull request by robrix**
This PR:

1. Corrects `Source` generation to produce multiline values, especially ones involving carriage returns as well as newlines.
2. Corrects the expectation of a `sourceLineRanges` property w.r.t. the number of lines produced to factor in carriage returns.
3. Labels the properties with 0/1/n lines, so we can see how many values were tested in each category.

Fixes #215.

### CODE REVIEW

File: test/Data/Source/Spec.hs [hunk 1]
```diff
@@ -25,11 +27,13 @@ prop desc f
 testTree :: Tasty.TestTree
 testTree = Tasty.testGroup "Data.Source"
   [ Tasty.testGroup "sourceLineRanges"
-    [ prop "produces 1 more range than there are newlines" $
-      \ source -> length (sourceLineRanges source) === succ (Text.count "\n" (toText source))
+    [ QC.testProperty "produces 1 more range than there are newlines" $
+      \ source -> QC.label (summarize source) $
+        length (sourceLineRanges source) QC.=== length (Text.splitOn "\r\n" (toText source) >>= Text.splitOn "\r" >>= Text.splitOn "\n")
```
Comment (robrix) [inline test/Data/Source/Spec.hs]: Ironically, this implementation is approximately as efficient as `newlineIndices` 😭

File: test/Data/Source/Spec.hs [hunk 2]
```diff
@@ -68,6 +72,10 @@ testTree = Tasty.testGroup "Data.Source"
     ]
 
   ]
+  where summarize src = case sourceLines src of
+          []  -> "empty"
+          [x] -> if nullSource x then "empty" else "single-line"
+          _   -> "multiple lines"
```
Comment (robrix) [inline test/Data/Source/Spec.hs]: We could do some more summarizing w.r.t. the specific line endings in use, but as we actually use `sourceLineRanges` almost exclusively in error reporting from `Assignment`, I didn’t figure it was worth the effort.

File: test/Generators.hs [hunk 3]
```diff
@@ -9,4 +9,4 @@ import qualified Data.Source
 import Data.Functor.Identity
 
 source :: (GenBase m ~ Identity, MonadGen m) => Hedgehog.Range Int -> m Data.Source.Source
-source r = Data.Source.fromUTF8 <$> Gen.utf8 r Gen.unicode
+source r = Data.Source.fromUTF8 <$> Gen.utf8 r (Gen.frequency [ (1, pure '\r'), (1, pure '\n'), (20, Gen.unicode) ])
```
Comment (robrix) [inline test/Generators.hs]: I changed the distribution for the `hedgehog` generator to match the `QuickCheck` one.

**Comment by robrix** [comment]
cc @importantshock

**Comment by patrickt** [comment]
> cc @importantshock

Whomst? :)

**Comment by robrix** [comment]
> > cc @importantshock
> 
> Whomst? :)

lol, I wondered why it wasn’t hyperlinked 😂

File: test/Data/Source/Spec.hs [hunk 4]
```diff
@@ -27,13 +26,15 @@ prop desc f
...
-      \ source -> QC.label (summarize source) $
-        foldMap (`slice` source) (sourceLineRanges source) QC.=== source
+    [ testProperty "produces 1 more range than there are newlines" $ property $ do
+        source <- forAll (Gen.source (Hedgehog.Range.linear 0 100))
+        label (summarize source)
+        (length (sourceLineRanges source) === length (Text.splitOn "\r\n" (toText source) >>= Text.splitOn "\r" >>= Text.splitOn "\n"))
+
+    , testProperty "produces exhaustive ranges" $ property $ do
+        source <- forAll (Gen.source (Hedgehog.Range.linear 0 100))
+        label (summarize source)
```
Comment (robrix) [inline test/Data/Source/Spec.hs]: Ohhhh so _that’s_ what I was doing wrong. I was expecting it to be scoped, and misread the type signature.

Comment (patrickt) [reply test/Data/Source/Spec.hs]: Yeah, I made the same mistake too! And to describe the error message that GHC gives (some nonsense about not being able to unify `m0` with `()` when the error is one of arity) as “disgustingly poor” is generous 😠

**Pull request merged**